### ISSUE TITLE
Allow upgrading to the latest available Elgg release directly with possibly ommiting minor/major releases inbetween

### ISSUE THREAD

**Issue by iionly** [Elgg/Elgg#10141]
It might have been less of a problem in the past with a new minor release getting available only every 2 or 3 years. But ever since new minor releases are published every 3 months the number of people who are not upgrading their site to the new minor release immediatelly has likely increased. If they decide to upgrade once in while chances have increased that there are one or more minor releases between their current Elgg version in use and the most recent Elgg version.

Another problem might occur with the LTS branch still being supported but only the most recent minor release of the next higher major Elgg version getting bug/security fixes. This could result (and if only by mistake) in a broken upgrade path, i.e. a change added to 1.12 and then merged in 2.2 but not in 2.0/2.1 could mean that you might not be able to upgrade from 1.12 to 2.2 if the upgrade from 1.12 to 2.0 and/or 2.1 fails.

The plan of a new major release once per year is also likely to result in people wanting to omitt a major release completely, e.g. upgrading from 1.12 to 3.0 directly.

So, I think it would really make sense to re-work the upgrade process to be able to upgrade directly to the most recent release avaiable at that time. I would think it would be a really nice to have feature for Elgg 3.0 - with supporting a direct upgrade from Elgg 1.12 and any 2.x version (excluding any older versions by an approach suggested in https://github.com/Elgg/Elgg/issues/10140).

- [ ] update docs saying 'latest minor to any minor in next major' instead of upgrading one minor at a time (https://github.com/Elgg/Elgg/issues/10141#issuecomment-319890241)

**Comment by mrclay** [comment]
I definitely appreciate this issue. What this would really mean in practical terms is that upgrade scripts and async upgrade operations would have to be runnable under any minor version. That's most likely the case already, at least for 2.x. As for upgrading to 3.0 from 1.12, it's too early to say, but I bet that would be rocky.

**Comment by iionly** [comment]
As I understand the current problem that might occur with leaping over minor releases is that some "initial stuff" is done only once (maybe CAN only be done once) at the beginning of the upgrade process (beg your pardon for my wording in lack of a deeper understanding of what exactly happens). When doing the upgrade step by step for every minor/major release in sequence the "initial stuff" already takes into account the modifications done by one of the upgrade scripts of a previous release and expected to be changed in a newer upgrade script of the later Elgg release.

Would it be possible to add some kind of "stopping points" in upgrade script that tells Elgg that it needs to restart the upgrade process at this point (allowing re-initializing the "initial stuff")? So, kind of an outer loop that initialized the upgrade, then runs the upgrade scripts in sequence until a stopping point is reached (jump out of the inner loop) or all upgrade scripts have been executed.

I guess "async" upgrades refers to long running upgrade scripts (e.g. comments upgrade) and not to some kind of upgrade scripts that could run in parallel to other scripts. The only solution I can think of how to make async upgrades still working flawless when ommitting Elgg releases might be to make ALL upgrade scripts async upgrades. In addition to adding "stopping points" this might require to also add "starting points" to the real, long running async upgrade scripts. By "starting point" I mean that they won't start if the site has not yet reach a specific Elgg version. With all upgrade stuff managed from the "pending upgrades" page this could look like

<Pending upgrade 1> (this could include one or several short-running script)
<Pending upgrade 2> (a long running script that requires one or more scripts of <Pending upgrade 1> already finished)
<Pending upgrade 3> (again one or more short-running scripts)
etc.

So, the scripts wouldn't necessrily have to work on any version. The only thing that needs to work is that it is made sure that the scripts are executed in a defined order.

**Comment by jdalsem** [comment]
Need to decide what we want with this for 3.0

**Comment by jdalsem** [comment]
You can only upgrade to a new major from the previous majors last minor version. 

This means 2.3 -> 3.0 
but not 2.0 -> 3.0
and not 1.12 -> 3.0

In the case of 1.12 you first need to go to 2.3 and then to 3.0

We need to explain this in the upgrade docs.

Be also advised that plugin developers also have their own upgrade path for their plugins. Advice is to always stay updated then this will not be a big issue.

**Comment by iionly** [comment]
Is it safe to upgrade from 1.12 to 2.3 directly? That would be an improvement already as the docs might still say "only upgrade one minor release at a time" since the days of the 1.x days. And if the same would work in future (e.g. 2.3 LTS to whatever 3.x and so on) it would make things a lot easier.

I don't think that 3rd party plugins upgrade path can have much relevance here anyway. It's Elgg core that should mark the pace - and if going directly to the latest minor release of the next major release would be possible I think it's quite likely that any 3rd party plugins will also work fine.

**Comment by jdalsem** [comment]
> Is it safe to upgrade from 1.12 to 2.3 directly?

yes this should be possible.



> That would be an improvement already as the docs might still say "only upgrade one minor release at a time" since the days of the 1.x days.

That made sense then as 'minor' releases in the 1.x series were actually major releases. As of 2.0 we are commited to semantic versioning so no major/breaking changes in minor releases.



> And if the same would work in future (e.g. 2.3 LTS to whatever 3.x and so on) it would make things a lot easier.

2.3 to 3.x should be possible.



> I don't think that 3rd party plugins upgrade path can have much relevance here anyway. It's Elgg core that should mark the pace - and if going directly to the latest minor release of the next major release would be possible I think it's quite likely that any 3rd party plugins will also work fine.

Disagree a bit on this. Plugin devs also apply versioning to their plugins and they can provide major updates and migration scripts in a different release scheme or speed. I have plugins with multiple major releases while elgg is just putting out minors, so checking out plugins could be important.

i created a todo on this issue to update the docs with the description of the upgrade path
